lahja/endpoint.py

```diff
@@ -355,8 +381,21 @@ def stop(self) -> None:
         self._running = False
         self._receiving_queue.put_nowait((TRANSPARENT_EVENT, None))
         self._internal_queue.put_nowait((TRANSPARENT_EVENT, None))
+        for coro in self._child_coros:
+            coro.cancel()
+        self._server.close()
+        os.unlink(self.ipc_path)
+
+    def __enter__(self) -> 'Endpoint':
+        return self
+
+    def __exit__(self,
+                 exc_type: Optional[Type[BaseException]],
+                 exc_value: Optional[BaseException],
+                 exc_tb: Optional[TracebackType]) -> None:
+        self.stop()
 
-    def broadcast(self, item: BaseEvent, config: Optional[BroadcastConfig] = None) -> None:
+    async def broadcast(self, item: BaseEvent, config: Optional[BroadcastConfig] = None) -> None:
```

Comment:
I believe making this `async` will result in `async` creep and maybe difficult to resolve situations within Trinity. Nothing specific comes to my mind but I *think*  we use `broadcast` from quite a bunch of synchronous methods which may be hard or awkward to turn into `async`.

Do we need the backpressure for this to work? Or could this be an opt-in thing resulting in another method `broadcast_nowait` maybe? I quite like that `broadcast` has worked in a fire-and-forget manner so far. What happens if we broadcast an event to many consumers and one of them (e.g. a third party plugin) consumes them too slowly because of some blocking code. Doesn't that slow down the event propagation for all other consumers as well?

Reply:
On a second thought, I guess the `broadcast_nowait` could simply be implemented as a `asyncio.ensure_future(...)` wrapping the call like you did further down in the benchmark. But if there's a legit need for it I think adding this as a method such as `broadcast_nowait` on the endpoint would make a lot of sense.

Reply:
I'm also :+1: on a `_nowait` option as I think @cburgdorf is correct about the difficulties we'll find in trinity.

Reply:
I was worried about this one too but I think making `broadcast` async is the right thing to do. `broadcast` sends data into a socket and that socket could genuinely back up.

>  Or could this be an opt-in thing resulting in another method broadcast_nowait maybe? 

If it's really necessary I guess we can add this, but this is a scary method! There are two ways to do it:
- `broadcast` before this PR would just write into the socket, this blocked the event loop whenever the socket filled up. (and also blocked the event loop until the remote process had confirmed receipt, I think this is why performance has improved)
- `ensure_future` is how I had originally written this PR, but it failed when I ran: `perf_benchmark.py --num-processes 3 --num-events 10000 --payload-bytes 1000000`. This test fired off a bunch of futures, each of them holding a MB of data, each of them waiting to be able to write to the socket. For reasons I don't quite understand the process ended up consuming ~26GB of memory and swapping wildly as it tried to send. This is an extreme example, but it convinced me that it's much safer for `broadcast` to block the caller.

> What happens if we broadcast an event to many consumers and one of them (e.g. a third party plugin) consumes them too slowly because of some blocking code. Doesn't that slow down the event propagation for all other consumers as well?

I think it usually won't. The `InboundConnection` reads messages out and throws them into the unbounded `_receiving_queue`, so a slow plugin would only cause memory bloat in it's own process. An exception to this would be a plugin which manages to block the event loop. That would eventually fill the socket's buffer and start blocking broadcast.

It might be possible to detect this situation and emit a warning? I don't have a good answer for you, but I lean towards poor performance being better than unbounded memory growth.

Reply:
>If it's really necessary I guess we can add this, but this is a scary method!

I think it's ok to discourage usage of `broadcast_nowait` in the docstring but I still prefer to have it when the alternative is to force uses to use `asyncio.ensure_future(...) or even `cast(None, asyncio.ensure_future(...))` when used in a `lambda`. Especially, because `broadcast_nowait` can also set `loop=self._loop` to ensure we broadcast on the right loop and not on the default loop.